tests/custom_object_test.py

```diff
@@ -0,0 +1,279 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl.testing import absltest, parameterized
+
+import numpy as np
+
+from jax import test_util as jtu
+import jax.numpy as jnp
+from jax import core, jit, lax, lazy, make_jaxpr
+from jax.interpreters import xla
+from jax.lib import xla_client
+xops = xla_client.ops
+
+from jax.config import config
+config.parse_flags_with_absl()
+
+# Define a sparse array data structure. The important feature here is that
+# it is a jaxpr object that is backed by two device buffers.
+class SparseArray:
+  """Simple sparse COO array data structure."""
+  def __init__(self, aval, data, indices):
+    self.aval = aval
+    self.shape = aval.shape
+    self.data = data
+    self.indices = indices
+
+  @property
+  def index_dtype(self):
+    return self.indices.dtype
+
+  @property
+  def dtype(self):
+    return self.data.dtype
+
+  @property
+  def nnz(self):
+    return self.data.shape[0]
+
+  def __repr__(self):
+    return repr(list((tuple(ind), d) for ind, d in zip(self.indices, self.data)))
+
+
+class AbstractSparseArray(core.ShapedArray):
+  __slots__ = ['index_dtype', 'nnz', 'data_aval', 'indices_aval']
+  _num_buffers = 2
+
+  def __init__(self, shape, dtype, index_dtype, nnz):
+    super(AbstractSparseArray, self).__init__(shape, dtype)
+    self.index_dtype = index_dtype
+    self.nnz = nnz
+    self.data_aval = core.ShapedArray((nnz,), dtype)
+    self.indices_aval = core.ShapedArray((nnz, len(shape)), index_dtype)
+
+  @core.aval_property
+  def data(self):
+    return sp_data_p.bind(self)
+
+  @core.aval_property
+  def indices(self):
+    return sp_indices_p.bind(self)
+
+def abstract_sparse_array(arr):
+  return AbstractSparseArray(arr.shape, arr.dtype, arr.index_dtype, arr.nnz)
+
+def sparse_array_result_handler(device, aval):
+  def build_sparse_array(data_buf, indices_buf):
+    data = xla.DeviceArray(aval.data_aval, device, lazy.array(aval.data_aval.shape), data_buf)
+    indices = xla.DeviceArray(aval.indices_aval, device, lazy.array(aval.indices_aval.shape), indices_buf)
+    return SparseArray(aval, data, indices)
+  return build_sparse_array
+
+def sparse_array_shape_handler(a):
+  return (
+    xla.xc.Shape.array_shape(a.data_aval.dtype, a.data_aval.shape),
+    xla.xc.Shape.array_shape(a.indices_aval.dtype, a.indices_aval.shape),
+  )
+
+def sparse_array_device_put_handler(a, device):
+  return (
+    xla.xb.get_device_backend(device).buffer_from_pyval(a.data, device),
+    xla.xb.get_device_backend(device).buffer_from_pyval(a.indices, device)
+  )
+
+core.pytype_aval_mappings[SparseArray] = abstract_sparse_array
```

Comment:
IIUC this test file will have a global side-effect on all these dicts that persists after the test has finished. That seems pretty harmless, but it'd be cleaner to do this registration in module-level (or class-level) setUp / tearDown functions (or methods) so as to make this test module more hermetic.

If you don't feel like making that change now, we could just leave a TODO here to follow up later.

Reply:
Added the TODO for now.